pennylane/fourier/advanced_spectrum.py

```diff
@@ -0,0 +1,365 @@
+        argnum (list[int]): Numerical indices for arguments
+        qnode (QNode): QNode to infer the ``encoding_args`` and ``argnum`` from
+            if both are ``None``
+    Returns:
+        OrderedDict[str, list[tuple]]: Ordered parameter index dictionary;
+            keys are argument names, values are index tuples for that argument
+            or an ``Ellipsis``
+        list[int]: Numerical indices for arguments
+
+    If ``encoding_args`` are passed, they take precedence over ``argnum``.
+    Passing a set with ``keys`` is an alias for ``{key: ... for key in keys}``.
+    If both, ``encoding_args`` and ``argnum`` are ``None``, all QNode arguments
+    of ``qnode`` that do not have a default value defined are included.
+    Of these arguments, all elements are included in ``encoding_args``
+    """
+    sig_pars = signature(qnode.func).parameters
+    arg_names = list(sig_pars.keys())
+    arg_names_no_def = [name for name, par in sig_pars.items() if par.default is par.empty]
+
+    if encoding_args is None:
+        if argnum is None:
+            encoding_args = OrderedDict((name, ...) for name in arg_names_no_def)
+            argnum = list(range(len(arg_names_no_def)))
+        elif np.isscalar(argnum):
+            encoding_args = OrderedDict({arg_names[argnum]: ...})
+            argnum = [argnum]
+        else:
+            encoding_args = OrderedDict((arg_names[num], ...) for num in argnum)
+    else:
+        requested_names = set(encoding_args)
+        if not all(name in arg_names for name in requested_names):
+            raise ValueError(
+                f"Not all names in {requested_names} are known. " f"Known arguments: {arg_names}"
+            )
+        # Selection of requested argument names from sorted names
+        if isinstance(encoding_args, set):
+            encoding_args = OrderedDict(
+                (name, ...) for name in arg_names if name in requested_names
+            )
+        else:
+            encoding_args = OrderedDict(
+                (name, encoding_args[name]) for name in arg_names if name in requested_names
+            )
+        argnum = [arg_names.index(name) for name in encoding_args]
+
+    return encoding_args, argnum
+
+
+def expand_multi_par(tape, depth=10):
+    """Expand a tape until it does not contain any multi-parameter gates, if possible.
+
+    Args:
+        tape (.QuantumTape): Tape to be expanded
+        depth (int): Maximum expansion depth
+
+    Returns
+        .QuantumTape: Expanded tape
+
+    """
+    if any(len(op.parameters) > 1 for op in tape.operations):
+        stopping_cond = lambda g: (
+            isinstance(g, qml.measure.MeasurementProcess) or len(g.parameters) <= 1
+        )
+        new_tape = tape.expand(depth=depth, stop_at=stopping_cond)
+        params = new_tape.get_parameters(trainable_only=False)
+        new_tape.trainable_params = qml.math.get_trainable_indices(params)
```

Comment:
Just curious - are these two lines required?

Reply:
I don't think they are required within the spectrum functionality, because we never use the `trainable_params` attribute.
Somehow it felt weird to leave them out because when borrowing this expansion function this might yield side effects.
However, probably cleaner to remove, now that I reconsider :+1:

Reply:
Hmm you could leave it there just commented out (with an explanation as to why they are commented out?)